test/common/network/connection_impl_test.cc

```diff
@@ -870,6 +887,173 @@ TEST_P(ConnectionImplTest, EmptyReadOnCloseTest) {
   disconnect(true);
 }
 
+// Test that a FlushWrite close immediately triggers a close after the write buffer is flushed.
+TEST_P(ConnectionImplTest, FlushWriteCloseTest) {
+  setUpBasicConnection();
+  connect();
+  // Set a very high timeout value to prevent flaking. We are testing the common case where the
+  // timeout does not trigger.
+  server_connection_->setDelayedCloseTimeout(std::chrono::milliseconds(50000));
+
+  std::shared_ptr<MockReadFilter> client_read_filter(new NiceMock<MockReadFilter>());
+  client_connection_->addReadFilter(client_read_filter);
+
+  NiceMockConnectionStats stats;
+  server_connection_->setConnectionStats(stats.toBufferStats());
+
+  Buffer::OwnedImpl data("data");
+  server_connection_->write(data, false);
+
+  // Server connection flushes the write and immediately closes the socket.
+  // There shouldn't be a read/close race here (see issue #2929), since the client is blocked on
+  // reading and the connection should close gracefully via FIN.
+  EXPECT_CALL(*client_read_filter, onData(BufferStringEqual("data"), false))
+      .Times(1)
+      .WillOnce(InvokeWithoutArgs([&]() -> FilterStatus {
+        dispatcher_->exit();
+        return FilterStatus::StopIteration;
+      }));
+  EXPECT_CALL(stats.delayed_close_timeouts_, inc()).Times(0);
+  EXPECT_CALL(client_callbacks_, onEvent(ConnectionEvent::RemoteClose)).Times(1);
+  EXPECT_CALL(server_callbacks_, onEvent(ConnectionEvent::LocalClose)).Times(1);
+  server_connection_->close(ConnectionCloseType::FlushWrite);
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
+}
+
+// Test that a FlushWrite close will trigger a timeout which closes the connection when the write
+// buffer is not flushed within the configured interval.
+TEST_P(ConnectionImplTest, FlushWriteCloseTimeoutTest) {
+  NiceMock<Event::MockDispatcher> dispatcher;
+  EXPECT_CALL(dispatcher.buffer_factory_, create_(_, _))
+      .WillRepeatedly(Invoke([](std::function<void()> below_low,
+                                std::function<void()> above_high) -> Buffer::Instance* {
+        return new Buffer::WatermarkBuffer(below_low, above_high);
+      }));
+
+  auto timer = new Event::MockTimer(&dispatcher);
+  EXPECT_CALL(*timer, enableTimer(_)).Times(1);
+  EXPECT_CALL(*timer, disableTimer()).Times(1);
+
+  auto file_event = new NiceMock<Event::MockFileEvent>;
+  EXPECT_CALL(dispatcher, createFileEvent_(0, _, _, _)).WillOnce(Return(file_event));
+
+  auto transport_socket = new NiceMock<MockTransportSocket>;
+  EXPECT_CALL(*transport_socket, canFlushClose()).WillOnce(Return(true));
+
+  std::unique_ptr<Network::ConnectionImpl> server_connection(new Network::ConnectionImpl(
+      dispatcher, std::make_unique<ConnectionSocketImpl>(0, nullptr, nullptr),
+      TransportSocketPtr(transport_socket), true));
+  server_connection->setDelayedCloseTimeout(std::chrono::milliseconds(500));
```

Comment:
Are we safe from flakes here?

Reply:
Yes. The timer's callback is manually invoked at the end of the test. The 'setDelayedCloseTimeout()'  call is required to enable creation of the timer, but the actual value does not matter (as long as it's >0). I've added a clarifying comment.